feat(web-analytics): creating insight calendar heatmap

jabahamondes opened the pull request:
## Problem

Calendar heatmap had good feedback and was asked to create an insight for it.

## Changes

In this PR we create the most basic insight that allows us to create a calendar insight for total events and unique users for any kind of filters.

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Manually and query runner with unit tests.


Different use cases tested:

<img width="1192" alt="image" src="https://github.com/user-attachments/assets/ef58ec29-55cf-4340-aac6-5dc971f056ca" />
<img width="1185" alt="image" src="https://github.com/user-attachments/assets/231934ec-acc5-47bb-825e-7779e97621f6" />
<img width="1179" alt="image" src="https://github.com/user-attachments/assets/e5ba03dc-5f0f-48b0-afb2-adc94631efbf" />
<img width="1179" alt="image" src="https://github.com/user-attachments/assets/26821fca-e7d7-4bca-baed-7b7d60ad6019" />
<img width="604" alt="image" src="https://github.com/user-attachments/assets/9b7680f7-7e3b-466f-98e7-053a7fb64a0b" />
<img width="670" alt="image" src="https://github.com/user-attachments/assets/3ac8ee0f-fad2-4e50-83b4-33844ff16408" />
<img width="584" alt="image" src="https://github.com/user-attachments/assets/2d0fa398-673b-4a3f-9071-c7bc70b3ba7e" />
<img width="1245" alt="image" src="https://github.com/user-attachments/assets/f0668777-9e09-4a00-8576-9432a081e94b" />
<img width="1191" alt="image" src="https://github.com/user-attachments/assets/0ded5993-00f2-40c5-bc0b-33b09535a95c" />
<img width="1199" alt="image" src="https://github.com/user-attachments/assets/ba4b781f-111e-435c-b5f4-6aa2a43ac53b" />

Commit:
feat(web-analytics): add loading state to EventsHeatMap component

- Introduced loading state handling using InsightLoadingState and InsightsWrapper.
- Updated useValues to include responseLoading and queryId for better loading management.

Commit:
feat(web-analytics): add Active Hours tab and update priority mapping

- Introduced ACTIVE_HOURS enum to TileId and ActiveHoursTab enum.
- Updated loadPriorityMap to include ACTIVE_HOURS with adjusted priorities for existing tiles.
- Added setActiveHoursTab action and state management in webAnalyticsLogic.
- Integrated Active Hours tab into the web analytics logic and UI.

Commit:
fix(web-analytics): update web active hours heatmap query for unique session events

- Refactored the query to use unique session events for aggregating hours and days.
- Adjusted the subquery to group by session_id and ensure correct timestamp handling.

Commit:
Moving InsightsWrapper to own place

Commit:
Merge branch 'master' into feat/making-query-handle-unique-users-for-activity-hours

Commit:
Removing extra space

jabahamondes committed:
Merge branch 'master' into feat/making-query-handle-unique-users-for-activity-hours

Commit:
Update tests

Commit:
Add active hours documentation and enhance EventsHeatMap logic

- Introduced detailed documentation for the active hours feature in webAnalyticsLogic.tsx, including usage notes and recommendations for time range selection.
- Updated EventsHeatMap component to incorporate week start day logic, improving the accuracy of the heatmap display.
- Refactored data processing functions to handle new aggregation structures and ensure proper rendering of heatmap cells with updated tooltip functionality.

Commit:
Merge branch 'master' into feat/improving-activty-hours

Commit:
Refactor heatmap query structure and components

- Introduced new HeatMapQueryResponse and HeatMapStructuredResult types to replace the previous WebActiveHoursHeatMap structures.
- Updated EventsHeatMap component to utilize WebActiveHoursHeatmap for rendering, enhancing the heatmap display logic.
- Removed deprecated WebActiveHoursHeatMap related code and tests, streamlining the overall query handling.
- Added utility functions for tooltip generation and font size adjustments based on table width.
- Improved data processing logic to accommodate new data structures and ensure accurate rendering of heatmap cells.

Commit:
Merge branch 'master' into chore/refactoring-events-heatmap-to-scale-to-more-queries

Commit:
fix(EventsHeatMap): simplify font size calculation logic

Refactored the font size setting in the EventsHeatMap component to directly call the thresholdFontSize function, removing unnecessary intermediate variable. This change streamlines the code and improves readability.

Commit:
fix(heatmap): Correct background color logic for min/max values in HeatMapCell
- Adjusted the order of conditions to ensure proper saturation for min and max values.

Commit:
- Changed matrix type from object to a 2D array for better performance and simplicity.
- Updated columns and rows aggregations to use arrays instead of objects, enhancing consistency across the data processing functions.

Commit:
naming mixed fixing

Commit:
refactor(heatmap): Rename WebActiveHoursHeatMapQuery to HeatMapQuery
- Updated schema and query references to use HeatMapQuery instead of WebActiveHoursHeatMapQuery for consistency.
- Introduced new HeatMapQuery definition in schema.json and updated related utility functions and components accordingly.
- Adjusted all instances in the codebase to reflect the new naming convention, ensuring seamless integration with existing features.

Commit:
handling negative module

Commit:
Merge branch 'master' into chore/refactoring-events-heatmap-to-scale-to-more-queries

Commit:
renaming

Commit:
fixing tests

Commit:
feat(web-analytics): add conversion goal to web analytics logic and update event expression handling in heatmap query runner

Commit:
feat(web-analytics): add conversion goal to web analytics logic and update event expression handling in heatmap query runner

Commit:
feat: working insight, created all the logic related, empty filters, query runner, etc

Commit:
Merge branch 'master' into feat/insight-calendar-heatmap

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

Commit:
Merge branch 'master' into feat/insight-calendar-heatmap

Commit:
renaming

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

robbie-c commented on posthog/hogql_queries/insights/trends/calendar_heatmap_query_runner.py:
```diff
@@ -0,0 +1,381 @@
+
+    def to_query(self) -> ast.SelectQuery:
+        # Use the heatmap query logic
+        placeholder = placeholderUniqueUsers if self.query.series[0].math == "dau" else placeholderAllUsers
```
I think "placeholder" is not quite the right term for this, maybe `queryUniqueUsers` etc

jabahamondes replied:
I think template is a better word for this case ✅  let me know if your ok with it

Commit:
adding tests

robbie-c commented on posthog/hogql_queries/insights/trends/calendar_heatmap_query_runner.py:
```diff
@@ -0,0 +1,381 @@
+    query.total
+FROM (
+    SELECT
+        uniqMap(map(concat(toString(toDayOfWeek(events.timestamp)), ',' ,toString(toHour(events.timestamp))), events.uuid)) as hoursAndDays,
```
Does
```
sumMap(map(concat(toString(toDayOfWeek(events.timestamp)), ',' ,toString(toHour(events.timestamp))), 1))
```
give the same result? Might be slightly more performant

jabahamondes replied:
Yes, it did! thanks

jabahamondes commented on frontend/src/queries/nodes/InsightViz/CalendarHeatmapFilters.tsx:
```diff
@@ -0,0 +1,45 @@
+                filters={filters}
+                setFilters={(payload: Partial<FilterType>): void => {
+                    updateQuerySource({
+                        series: actionsAndEventsToSeries(payload as any, true, mathAvailability),
```
It's expected

jabahamondes replied:
Mm necessary due to how it's implemented

jabahamondes commented on frontend/src/queries/examples.ts:
```diff
@@ -206,6 +207,16 @@ const InsightTrendsQuery: TrendsQuery = {
 
+const InsightCalendarHeatmapQuery: CalendarHeatmapQuery = {
+    kind: NodeKind.CalendarHeatmapQuery,
+    properties: [],
```
It's expected

jabahamondes commented on frontend/src/queries/nodes/InsightQuery/defaults.ts:
```diff
@@ -25,6 +26,18 @@ export const trendsQueryDefault: TrendsQuery = {
     trendsFilter: {},
 }
 
+export const calendarHeatmapQueryDefault: CalendarHeatmapQuery = {
+    kind: NodeKind.CalendarHeatmapQuery,
+    series: [
+        {
+            kind: NodeKind.EventsNode,
+            name: '$pageview',
+            event: '$pageview',
+            math: BaseMathType.TotalCount,
+        },
+    ],
+}
```
Not necessary but still adding it

robbie-c commented on posthog/hogql_queries/insights/trends/calendar_heatmap_query_runner.py:
```diff
@@ -0,0 +1,381 @@
+            return "String"
+
+    # TODO: Move this to posthog/hogql_queries/legacy_compatibility/query_to_filter.py
+    def _query_to_filter(self) -> dict[str, Any]:
```
Do you need this? (maybe you do, I'm not super familiar with how these filters work)

jabahamondes replied:
No we don't need it🙏

robbie-c commented on posthog/hogql_queries/insights/trends/calendar_heatmap_query_runner.py:
```diff
@@ -0,0 +1,381 @@
+    query.total
+FROM (
+    SELECT
+        uniqMap(map(concat(toString(toDayOfWeek(uniqueSessionEvents.timestamp)), ',' ,toString(toHour(uniqueSessionEvents.timestamp))), uniqueSessionEvents.person_id)) as hoursAndDays,
```
Ah nice, concat is a more flexible way to do it than multiplying by 100 or however we handled it in London :)

jabahamondes replied:
Mm I'm not sure if it would work in this case bc I'm not grouping by anything, but `concatWithSeparator` is better than may concat.

robbie-c reviewed (commented):
Left a few comments, will have a deeper review tomorrow morning :)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

jabahamondes commented on frontend/src/scenes/trends/mathsLogic.tsx:
```diff
@@ -70,6 +77,37 @@ export const FUNNEL_MATH_DEFINITIONS: Record<FunnelMathType, MathDefinition> = {
     },
 }
 
+export const MAPA_CALOR_MATH_DEFINITIONS: Record<CalendarHeatmapMathType, MathDefinition> = {
```
Thanks, add an unsaved file 🤦  sorry

jabahamondes commented on posthog/schema.py:
```diff
@@ -413,6 +413,18 @@ class StatusItem(BaseModel):
     value: str
 
 
+class CalendarHeatmapFilter(BaseModel):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    dummy: Optional[str] = None
+
```
It was in purpose, but open to remove it

jabahamondes commented on posthog/hogql_queries/insights/trends/calendar_heatmap_query_runner.py:
```diff
@@ -0,0 +1,381 @@
+        # Process day-hour combinations
+        for i in range(len(hours_and_days_keys)):
+            key = hours_and_days_keys[i]
+            day, hour = map(int, key.split(","))
+            total = int(hours_and_days_values[i])
```
I shouldn't fail, since we use the same separator, I will leave a comment.

jabahamondes replied:
Endup updating with the same separator

Commit:
feat(insight): PR's comments: add calendarHeatmapFilter to defaults and update related types

- Introduced  in .
- Exported  type for better type safety in filters.
- Updated  to use the new type for .
- Enhanced  description for clarity.
- Renamed constants in  for consistency and clarity.

Commit:
Merge branch 'feat/insight-calendar-heatmap' of github.com:PostHog/posthog into feat/insight-calendar-heatmap

jabahamondes committed:
Refactor: Remove EventsHeatMapQuery and fixing bugs  (#32129)

Co-authored-by: Javier Bahamondes <javierbahamondes@Javiers-MacBook-Pro.local>

Commit:
Merge branch 'feat/insight-calendar-heatmap' of github.com:PostHog/posthog into feat/insight-calendar-heatmap

Commit:
re updating pngs and ts error

Commit:
reverting pngs

Commit:
fixing CI

Commit:
fixing CI

Commit:
resolving flaky test

lricoy commented on frontend/src/products.tsx:
```diff
@@ -337,6 +337,11 @@ export const getTreeItemsNew = (): FileSystemImport[] => [
     { path: `Early access feature`, type: 'early_access_feature', href: urls.earlyAccessFeature('new') },
     { path: `Experiment`, type: 'experiment', href: urls.experiment('new') },
     { path: `Feature flag`, type: 'feature_flag', href: urls.featureFlag('new') },
+    {
+        path: `Insight/Calendar Heatmap`,
+        type: 'insight',
+        href: urls.insightNew({ type: InsightType.CALENDAR_HEATMAP }),
+    },
```
FYI: This is actually a generated file. It is a recent change and I am not sure if it was just commited or you had to update it manually, so commenting only for the sake of sharing this.

You've already added it to the `manifest.tsx` so I guess all is good :D https://github.com/PostHog/posthog/blob/fe5e6012b6fb48e5163215e0752c433dc63f217e/products/product_analytics/manifest.tsx#L109-L113

lricoy commented on frontend/src/queries/Query/Query.tsx:
```diff
@@ -175,7 +175,7 @@ export function Query<Q extends Node>(props: QueryProps<Q>): JSX.Element | null
         component = <HogDebug query={query} setQuery={setQuery as (query: any) => void} queryKey={String(uniqueKey)} />
-    } else if (isEventsHeatMapQuery(query)) {
+    } else if (isCalendarHeatmapQuery(query)) {
         component = <WebActiveHoursHeatmap query={query} context={queryContext} cachedResults={props.cachedResults} />
```
Maybe you want to update the component name as well, but no need to do it on this PR. (I did not check the other one so far, so this could be done already)

lricoy commented on frontend/src/queries/schema/schema-general.ts:
```diff
@@ -1011,6 +1016,11 @@ export type TrendsFilter = {
     goalLines?: GoalLine[]
 }
 
+export type CalendarHeatmapFilter = {
+    // Reserved for future filter properties
+    dummy?: string
+}
```
Is there a problem to leave it as a empty type?

lricoy commented on posthog/hogql_queries/insights/trends/calendar_heatmap_query_runner.py:
```diff
@@ -0,0 +1,334 @@
+    mapKeys(query.days) as daysKeys,
+    mapValues(query.days) as daysValues,
+    query.total
+FROM (
+    SELECT
+        uniqMap(map(concatWithSeparator({separator},toString(toDayOfWeek(uniqueSessionEvents.timestamp)),toString(toHour(uniqueSessionEvents.timestamp))), uniqueSessionEvents.person_id)) as hoursAndDays,
+        uniqMap(map(toHour(uniqueSessionEvents.timestamp), uniqueSessionEvents.person_id)) as hours,
+        uniqMap(map(toDayOfWeek(uniqueSessionEvents.timestamp), uniqueSessionEvents.person_id)) as days,
+        uniq(person_id) as total
+    FROM (
+        SELECT
+            any(events.person_id) as person_id,
+            session.session_id as session_id,
+            min(session.$start_timestamp) as timestamp
+        FROM events
+        WHERE and(
+            {event_expr},
+            {all_properties},
+            {test_account_filters}
+        )
+        GROUP BY session_id
+    ) as uniqueSessionEvents
+    WHERE {current_period}
+) as query
```
I think the session's join here will only hold true for the context of WebAnalytics/SessionReplay. It is flipping the problem you had for the WebOverview the other way around: it won't match the `math` for `unique users` using only events like others insights will do.

e.g: if people have a "Big Number" tile in a dashboard for the count of unique users, the `query.total` here won't match.

IMHO it is better to count using only events for the time being to keep it more performant and simple and maybe add a `countUsingSessions` options to keep the behaviour for web analytics on another PR.

robbie-c replied:
+1 to this

lricoy commented on posthog/hogql_queries/insights/trends/calendar_heatmap_query_runner.py:
```diff
@@ -0,0 +1,334 @@
+"""
+
+
+class CalendarHeatmapQueryRunner(QueryRunner):
```
This lgtm, but you may want to check later if the ProductAnalytics folks want to make the resulting query a bit more similar to the ones they use, like using `_get_date_subqueries`. That said, I am in no position to check if that is recommended or not; I was just checking the implementation differences between our queries on the `WebAnalyticsQueryRunner` and the `TrendsQueryBuilder` for instance.

lricoy commented on playwright/e2e/product-analytics/retention.spec.ts:
```diff
@@ -20,8 +20,8 @@ test.describe('Retention', () => {
         await page.fill('[data-attr=taxonomic-filter-searchfield]', 'is_demo')
-        await page.click('[data-attr=taxonomic-tab-person_properties]')
-        await page.click('[data-attr=prop-filter-person_properties-0]')
+        await page.click('[data-attr=taxonomic-tab-event_properties]')
+        await page.click('[data-attr=prop-filter-event_properties-0]')
```
Are those intended for fixing the CI or something?

jabahamondes replied:
Yes! I run the test locally and was flaky, when followed the tests steps couldn’t get through the steps unless I change the selector

lricoy commented on frontend/src/types.ts:
```diff
@@ -2546,6 +2547,11 @@ export interface PathsFilterType extends FilterType {
     path_dropoff_key?: string // Paths People Dropoff Key
 }
 
+export interface CalendarHeatmapFilterType extends FilterType {
+    // Reserved for future filter properties
+    dummy?: string
+}
```
Same here, if it doesn't let you keep an empty type for the linter, I think it is best to add a linter exception than to keep `dummy`

lricoy reviewed (commented):
It worked for me running locally and it is behind a FF, so LGTM.

Left some comments but none is required.

lricoy reviewed (approved):
Solid work @jabahamondes , congrats. 

I think you should wait for Robbies' review but since the PR is big, I will leave the approve to unblock :)

robbie-c reviewed (approved):
LGTM to merge as-is!

There's some potential follow-up work about how web analytics groups by sessions but product analytics doesn't, but IMO that can be done in a follow-up PR

jabahamondes commented:
Will do a follow up PR with the comments ✅
